syntax = "proto3";
package com.spotify.useraccount.v1;

import "google/protobuf/timestamp.proto";
import "google/protobuf/wrappers.proto";

enum AccountIdentifierType {
    UNKNOWN = 0;
    USER_ID = 1;
    USERNAME = 2;
    FACEBOOK_UID = 3;
    EMAIL = 4;
    PHONE_NUMBER = 5;
    APPLE_ID = 6;
    SAMSUNG_ID = 7;
    GOOGLE_ID = 8;
    AMAZON_ID = 9;
    DISCORD_ID = 10;
}

message AccountIdentifier {
    oneof value {
        string username = 1;
        string user_id = 2;
        string facebook_uid = 3;
        string email = 4;
        string phone_number = 5;
        string apple_id = 6;
        string samsung_id = 7;
        string google_id = 8;
        string amazon_id = 9;
        string discord_id = 10;
    }
}


// Uses StringValue instead of string to be able to distinguish between
// value set to default (empty string) and value not set.
message AccountIdentifiers {
    google.protobuf.StringValue username = 1;
    google.protobuf.StringValue user_id = 2;
    google.protobuf.StringValue facebook_uid = 3;
    google.protobuf.StringValue email = 4;
    google.protobuf.StringValue phone_number = 5;
    google.protobuf.StringValue apple_id = 6;
    google.protobuf.StringValue samsung_id = 7;
    google.protobuf.StringValue google_id = 8;
    google.protobuf.StringValue amazon_id = 9;
    google.protobuf.StringValue discord_id = 10;
}

message AccountAttribute {
    oneof value {
        bool bool_value = 2;
        int64 long_value = 3;
        string string_value = 4;
        google.protobuf.Timestamp timestamp_value = 5;
    }
    bool client_visible = 6;
}

message Account {
    AccountIdentifiers identifiers = 1;
    map<string, AccountAttribute> attributes = 2;
}

message GetAccountRequest {
    AccountIdentifier identifier = 1;
    repeated AccountIdentifierType requested_identifiers = 2;
    repeated string requested_attributes = 3;
}

message GetAccountResponse {
    Account account = 1;
}

service UserAccount {
    rpc GetAccount(GetAccountRequest) returns (GetAccountResponse) {}
}
